### ISSUE TITLE
Support the most esoteric setup.py files

### ISSUE THREAD

**Issue by ShadowJonathan** [dependabot/dependabot-core#3202]
While working on trying to figure out why dependabot doesn't work for [synapse](https://github.com/matrix-org/synapse)'s [`setup.py`](https://github.com/matrix-org/synapse/blob/2566dc57ce69b1a141014cc7231f84ea6d3f3ea7/setup.py) file, I came across the python `setup.py` helper, which in my opinion takes a *huge* amount of liberties to assume how `setup.py` files look like.

This helper tries to outsmart python itself by manually regex-replacing(!) contents of the file, before blindly `exec`-ing it with `locals()` and `globals()`.

This can be simplified instead by using `multiprocessing` (to isolate side-effects), then run a script that replaces `setuptools`'s `setup` by making it send the passed requires and extras to the parent process, and then springboarding into the `setup.py` script (after setting the right working directory).

The only reason why i could see trying to outsmart python in regards to executing *what* in files would be to avoid malicious behaviour, that can be wrapped and isolated with containers, but that's not the case here, it's avoidance to dirty the output and such, *that* can be wrapped in a subprocess.

**Comment by jurre** [comment]
Are you referring to [this](https://github.com/dependabot/dependabot-core/blob/ce711fc43dd8c5ed3e978366780498ee5d1f1bc4/python/helpers/lib/parser.py#L59) helper?

**Comment by ShadowJonathan** [comment]
Yeah.

Also, i apologise for the tone in the initial message, I was still a bit weirded out at that point, and so i could've put my concerns down in a different manner.

**Comment by jeffwidman** [comment]
If it's something you'd like to see fixed and willing to submit a PR, I'm happy to give you pointers on how to debug further to pinpoint what code needs to change within `dependabot-core`.

However, I doubt the core team will spend much time on this as the python ecosystem seems to be moving away from `setup.py` in favor of `pyproject.toml`.

**Comment by ShadowJonathan** [comment]
I agree that with `pyproject.toml` being a viable alternative to `setup.py` in most (if not all) aspects, and with it being the successor to `setup.py`, that it being out of scope for dependabot is understandable.

I'll close this based on that. Feel free to re-open if this feature request still exists, but I have no longer any interest in it.
